simulation/response_functions.py

```diff
@@ -138,4 +137,137 @@ def gaussian_lateral(
+      wavelength: float denoting wavelength in meters
+
+    Returns:
+      1. Cross-section of beam: `np.ndarray` of shape `[length]` x `[length]` which stores u_J(x, z)
+      2. Full beam: `np.ndarray` of shape `[length]` x `[length]` x `[length]` which stores u_J(x, z)
+    """
+
+    if (beam.ndim != 3) and (beam.ndim != 4):
+        raise ValueError("dimension of `beam` must be 3 or 4, but got dimension {}".format(beam.ndim))
+
+    if (beam.ndim == 3):
+        x = beam[:, :, 0];
+        z = beam[:, :, 1];
+    else:
+        x = beam[:, :, :, 0];
+        z = beam[:, :, :, 1];
+
+    raleigh_length = np.pi * waist_radius ** 2 / wavelength
+    waist_z = waist_radius * np.sqrt(1 + (z / raleigh_length) ** 2)
+    complex_beam = z + 1j * raleigh_length
+    complex_beam_conjugate = z - 1j * raleigh_length
+
+    norm_constant = np.sqrt(np.sqrt(2/np.pi)/ np.exp2(degree) / np.math.factorial(degree) / waist_radius)
+    norm_z = np.sqrt(1j * waist_radius / complex_beam)
+    phase_shift = np.power(-complex_beam_conjugate / complex_beam, degree/2)
+    hermite = hermite_val(np.sqrt(2) * x / waist_z, degree)
+    amplitude_decay = np.exp(-1j * np.pi / wavelength * (x**2) / complex_beam)
+
+    return norm_constant * norm_z * phase_shift * hermite * amplitude_decay
+
+
+#TODO include different lengths and grid sizes in all three directions
+def hermite_gaussian_mode(
+    length: int,
+    wavelength: float,
+    dz: float,
+    L: int,
+    M: int = 0,
+    numerical_aperature: float = .125,
+    amplitude: float = 1.,
+):
+    """Computes complex E(x, y, z) over the volume of a `length * dz` x `length * dz` x `length * dz` cube
+    centered at the origin (measured in meters)
+
+    Reference: https://en.wikipedia.org/wiki/Gaussian_beam#Hermite-Gaussian_modes
+
+    Args:
+      length: int denoting number of pixels
+      wavelength: float denoting wavelength in meters.
+      dz: float denoting grid size in meters.
+      L, M: ints denoting (L, M) order of hermite-gaussian
+      numerical_aperture: float denoting NA of collecting apparatus.
+      amplitude: float denoting amplitude of beam (more specifically, E(0, 0, 0))
+
+    Returns:
+      `np.ndarray` of shape `[length]` x `[length]` x `[length]` which stores E(x, y, z)
+    """
+
+    if length % 2 != 1:
+      raise ValueError("`length` must be odd, but got {}".format(length))
+
+    center = length // 2 * dz
+    """ Creates an evenly spaced `[length]` x `[length]` x `[length]` 3-d grid with each element as [x, y, z] coordinates
+
+    Note: beam is of shape `[length]` x `[length]` x `[length]` x 3 as it is a 3-d grid which stores another 1-d grid
+    Reference: https://stackoverflow.com/questions/32208359/is-there-a-multi-dimensional-version-of-arange-linspace-in-numpy
+
+    """
+    beam = np.mgrid[-center:center:length*1j, -center:center:length*1j, -center:center:length*1j].reshape(3,-1).T.reshape(length, length, length, -1)
```

Comment:
This functionality is general (making a 3D grid so I would factor it into its own module as this may be useful elsewhere in the codebase.  Feel free to just put it in this module for now, though.  

When doing so please also separate each of the steps and more documentation to help the reader.

Thanks!